eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/jpa/LegacyJpaEventStorageEngine.java

```diff
@@ -0,0 +1,653 @@
+import org.axonframework.messaging.Message;
+import org.axonframework.messaging.MessageNameResolver;
+import org.axonframework.messaging.MessageStream;
+import org.axonframework.serialization.Serializer;
+import org.axonframework.serialization.upcasting.event.EventUpcaster;
+import org.axonframework.serialization.upcasting.event.NoOpEventUpcaster;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Spliterators;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.UnaryOperator;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static org.axonframework.common.BuilderUtils.*;
+import static org.axonframework.common.ObjectUtils.getOrDefault;
+import static org.axonframework.eventhandling.EventUtils.upcastAndDeserializeTrackedEvents;
+import static org.axonframework.eventsourcing.EventStreamUtils.upcastAndDeserializeDomainEvents;
+import static org.axonframework.eventsourcing.eventstore.LegacyAggregateBasedEventStorageEngineUtils.*;
+import static org.axonframework.eventsourcing.eventstore.LegacyAggregateBasedEventStorageEngineUtils.resolveAggregateIdentifier;
+
+
+/**
+ * EventStorageEngine implementation that uses JPA to store and fetch events.
+ * <p>
+ * By default, the payload of events is stored as a serialized blob of bytes. Other columns are used to store meta-data
+ * that allow quick finding of DomainEvents for a specific aggregate in the correct order.
+ *
+ * @author Mateusz Nowak
+ * @since 5.0.0
+ */
+public class LegacyJpaEventStorageEngine implements AsyncEventStorageEngine {
+
+    private static final Logger logger = LoggerFactory.getLogger(LegacyJpaEventStorageEngine.class);
+    private static final String DOMAIN_EVENT_ENTRY_ENTITY_NAME = DomainEventEntry.class.getSimpleName();
+
+    private final EntityManagerProvider entityManagerProvider;
+    private final TransactionManager transactionManager;
+    private final Serializer eventSerializer;
+    private final EventUpcaster upcasterChain;
+    private final PersistenceExceptionResolver persistenceExceptionResolver;
+    private final boolean explicitFlush;
+
+    private final MessageNameResolver messageNameResolver; // todo: use while upcasting implementing MessageName
+    private final LegacyJpaEventStorageOperations legacyJpaOperations;
+    private final BatchingEventStorageOperations batchingOperations;
+    private final GapAwareTrackingTokenOperations tokenOperations;
+
+    public LegacyJpaEventStorageEngine(
+            @javax.annotation.Nonnull EntityManagerProvider entityManagerProvider,
```

Comment:
Looks like the wrong Nonnull annotations are being used here. Please use the ones in  `jakarta.annotation` (as imported in the class)